src/appleseed/renderer/modeling/bsdf/hairbsdf.cpp

```diff
@@ -0,0 +1,840 @@
+            float sin_gamma_t = values->m_precomputed.m_h / eta_p;
+            float cos_gamma_t = std::sqrt(std::max(0.0f, 1.0f - sin_gamma_t * sin_gamma_t));
+            float gamma_t = std::asin(clamp(sin_gamma_t, -1.0f, 1.0f));
+            float theta_o = std::asin(clamp(sin_theta_o, -1.0f, 1.0f));
+
+            // Deviation of outgoing angle due to cuticular scales
+            float theta_o_R = theta_o - 2.0f * deg_to_rad(values->m_alpha);
+            float theta_o_TT = theta_o + deg_to_rad(values->m_alpha);
+            float theta_o_TRT = theta_o + 4.0f * deg_to_rad(values->m_alpha);
+
+            // Compute Attenuation for the single path through the hair cylinder
+            Spectrum T = exp(-values->m_precomputed.m_sigma_a * ((2.f * cos_gamma_t) / cos_theta_t));
+
+            float phi = phi_i - phi_o;
+            std::array<float, 4> ap_pdf = attenuation_pdf(cos_theta_o, values->m_eta, values->m_precomputed.m_h, T);
+
+            float bsdf_pdf = 0.f;
+            // RR contribution of scattering
+            bsdf_pdf +=
+                longitudinal(cos_theta_i, std::abs(std::cos(theta_o_R)),
+                    sin_theta_i, std::sin(theta_o_R), values->m_precomputed.m_v) *
+                ap_pdf[0] * azimuthal(phi, 0, values->m_precomputed.m_s, gamma_o, gamma_t);
+
+            // TT contribution of scattering
+            bsdf_pdf +=
+                longitudinal(cos_theta_i, std::abs(std::cos(theta_o_TT)), sin_theta_i,
+                    std::sin(theta_o_TT), values->m_precomputed.m_v * .25f) *
+                ap_pdf[1] * azimuthal(phi, 1, values->m_precomputed.m_s, gamma_o, gamma_t);
+
+            // TRT contribution of scattering
+            bsdf_pdf +=
+                longitudinal(cos_theta_i, std::abs(std::cos(theta_o_TRT)), sin_theta_i,
+                    std::sin(theta_o_TRT), values->m_precomputed.m_v * 4.f) *
+                ap_pdf[2] * azimuthal(phi, 2, values->m_precomputed.m_s, gamma_o, gamma_t);
+
+            // TRRT+ contribution of scattering
+            bsdf_pdf += longitudinal(cos_theta_i, cos_theta_o, sin_theta_i, sin_theta_o, values->m_precomputed.m_v * 4.f) *
+                ap_pdf[3] / (2.0f * pi);
+
+            return bsdf_pdf;
+        }
+
+        //
+        // Computes absorption coefficient based on eumelanin and pheomelanin concentrations.
+        //
+
+        Spectrum sigma_a_from_melanin(float melanin_conc, float melanin_ratio) const
+        {
+            float melanin = -std::log(std::max(1.f - melanin_conc, 0.0001f));
+            float ce = melanin * (1.f - melanin_ratio);
+            float cp = melanin * melanin_ratio;
+            Vector3f eumelanin_sigma_a = Vector3f(0.506f, 0.841f, 1.653f);
+            Vector3f pheomelanin_sigma_a = Vector3f(0.343f, 0.733f, 1.924f);
+            Vector3f sigma_a = (ce * eumelanin_sigma_a + cp * pheomelanin_sigma_a);
+            Spectrum ret(0.f);
```

Comment:
Don't initialize the spectrum if you immediately set it to another value afterward.